src/hotspot/cpu/x86/peephole_x86_64.cpp

```diff
@@ -235,6 +236,132 @@ bool Peephole::test_may_remove(Block* block, int block_index, PhaseCFG* cfg_, Ph
   return false;
 }
 
+// This function removes redundant lea instructions that result from chained dereferences that
+// match to leaPCompressedOopOffset, leaP8Narow, or leaP32Narrow. This happens for ideal graphs
+// of the form LoadN -> DecodeN -> AddP. Matching with any leaP* rule consumes both the AddP and
+// the DecodeN. However, after matching the DecodeN is added back as the base for the leaP*,
+// which is nessecary if the oop derived by the leaP* gets added to an OopMap, because OopMaps
+// cannot contain derived oops with narrow oops as a base.
+// This results in the following graph after matching:
+//  LoadN
+//  |   \
+//  | decodeHeapOop_not_null
+//  |   /       \
+//  leaP*    MachProj (leaf)
+// The decode_heap_oop_not_null will emit a lea with an unused result if the derived oop does
+// not end up in an OopMap.
+// This peephole recognizes graphs of the shape as shown above, ensures that the result of the
+// decode is only used by the derived oop and removes that decode if this is the case. Futher,
+// multipe leaP*s can have the same decode as their base. This peephole will remove the decode
+// if all leaP*s and the decode share the same parent.
+// Additionally, if the register allocator spills the result of the LoadN we can get such a graph:
+//               LoadN
+//                 |
+//        DefinitionSpillCopy
+//           /           \
+// MemToRegSpillCopy   MemToRegSpillCopy
+//           |           /
+//           | decodeHeapOop_not_null
+//           |   /              \
+//           leaP*          MachProj (leaf)
+// In this case where te common parent of the leaP* and the decode is one MemToRegSpill Copy
+// away, this peephole can als recognize the decode as redundant and also remove the spill copy
```

Comment:
Typo: `als`

Reply:
Fixed in [fb728f9](https://github.com/openjdk/jdk/pull/25471/commits/fb728f925442729b111fbe2b2c3cd57e3ed659c0)